token/core/fabtoken/driver/base.go

```diff
@@ -38,80 +40,65 @@ func (d *base) DefaultValidator(pp driver.PublicParameters) (driver.Validator, e
 
 func (d *base) newWalletService(
 	tmsConfig driver.Config,
-	binder common2.NetworkBinderService,
+	binder driver2.NetworkBinderService,
 	storageProvider identity.StorageProvider,
-	qe common.TokenVault,
+	qe driver.QueryEngine,
 	logger logging.Logger,
 	fscIdentity driver.Identity,
 	networkDefaultIdentity driver.Identity,
 	pp driver.PublicParameters,
 	ignoreRemote bool,
-) (*common.WalletService, error) {
+) (*wallet.Service, error) {
 	tmsID := tmsConfig.ID()
 
-	// Prepare roles
-	roles := identity.NewRoles()
 	deserializerManager := sig.NewMultiplexDeserializer()
 	identityDB, err := storageProvider.OpenIdentityDB(tmsID)
 	if err != nil {
 		return nil, errors.Wrapf(err, "failed to open identity db for tms [%s]", tmsID)
 	}
 	sigService := sig.NewService(deserializerManager, identityDB)
-	ip := identity.NewProvider(identityDB, sigService, binder, NewEIDRHDeserializer())
-	identityConfig, err := config2.NewIdentityConfig(tmsConfig)
+	identityProvider := identity.NewProvider(logger.Named("identity"), identityDB, sigService, binder, NewEIDRHDeserializer())
+	identityConfig, err := config.NewIdentityConfig(tmsConfig)
 	if err != nil {
 		return nil, errors.WithMessage(err, "failed to create identity config")
 	}
 
-	roleFactory := msp.NewRoleFactory(
-		logger,
-		tmsID,
-		identityConfig,         // config
-		fscIdentity,            // FSC identity
-		networkDefaultIdentity, // network default identity
-		ip,
-		ip, // signer service
-		ip, // endpoint service
-		storageProvider,
-		deserializerManager,
-		ignoreRemote,
-	)
-	role, err := roleFactory.NewWrappedX509(driver.OwnerRole, ignoreRemote)
+	// Prepare roles
+	roleFactory := role.NewFactory(logger, tmsID, identityConfig, fscIdentity, networkDefaultIdentity, identityProvider, identityProvider, identityProvider, storageProvider, deserializerManager)
+	role, err := roleFactory.NewRole(identity.OwnerRole, false, nil, x509.NewKeyManagerProvider(identityConfig, msp.RoleToMSPID[identity.OwnerRole], identityProvider, ignoreRemote))
 	if err != nil {
 		return nil, errors.WithMessage(err, "failed to create owner role")
 	}
-	roles.Register(driver.OwnerRole, role)
-	role, err = roleFactory.NewX509(driver.IssuerRole, pp.Issuers()...)
+	roles := wallet.NewRoles()
+	roles.Register(identity.OwnerRole, role)
+	role, err = roleFactory.NewRole(identity.IssuerRole, false, pp.Issuers(), x509.NewKeyManagerProvider(identityConfig, msp.RoleToMSPID[identity.IssuerRole], identityProvider, ignoreRemote))
```

Comment:
To give a better idea of what it is about, we can mention in the documentation that each wallet service has 4 roles.

Reply:
4 are the default roles. One can introduce more roles as needed.